### PR TITLE
383 fe view individual blog posts

### PR DESCRIPTION
**Pull request by mitkapanarin**

### CODE REVIEW

**Comment by edichoska** [comment]
@mitkapanarin Just a reminder: We are creating branches from staging, and we are merging the PRs to staging. So when you create a PR you should choose the base to which you are merging to be staging.

File: app/content-panel/blogs/[id]/BlogDetailsPage.module.scss [hunk 1]
```diff
@@ -50,6 +50,7 @@
     font-weight: bold;
     font-size: xx-large;
     width: 100%;
+    color: #a5a4a4;
```
Comment (edichoska) [inline app/content-panel/blogs/[id]/BlogDetailsPage.module.scss:53]: use colors that are defined in _colors.scss
same for typography, we have all the styles defined in _typography.scss

File: app/content-panel/blogs/[id]/page.tsx [hunk 2]
```diff
@@ -11,7 +11,7 @@ const BlogDetailsPage = ({ params }: { params: { id: string } }) => {
   const [slug, setSlug] = useState('N/A');
```
Comment (edichoska) [inline app/content-panel/blogs/[id]/page.tsx:11]: we are not editing the slug.
slug is the part of a URL that identifies a particular page on a website.
In next.js we use for creating dynamic routes, more information here:https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes

File: app/content-panel/blogs/[id]/page.tsx [hunk 3]
```diff
@@ -11,7 +11,7 @@ const BlogDetailsPage = ({ params }: { params: { id: string } }) => {
...
   const [image, setImage] = useState('');
   const [author, setAuthor] = useState({ first_name: 'N/A', last_name: 'N/A' });
-  const [publish_date, setPublishDate] = useState('N/A');
+  const [publishDate, setPublishDate] = useState('N/A');
```
Comment (edichoska) [inline app/content-panel/blogs/[id]/page.tsx:14]: all this fields/states will have the state for that blogpost we have fetched from the api
Here we are viewing and later editing and deleting existing blog posts.
For a blog post to be created, most of this fields would be required.

**Review by edichoska — CHANGES_REQUESTED** [review]

File: components/editor/TiptapEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  onChange,
+}: {
+  content: string;
+  editable: boolean;
+  onChange?: (content: string) => void;
+}) => {
+  const editorInstance = useEditor({
+    extensions: [
+      StarterKit,
+      Highlight,
```
Comment (edichoska) [inline components/editor/TiptapEditor.tsx:20]: Can you check if we can add markup?
or code blocks

File: app/content-panel/blogs/[id]/page.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+const BlogDetailsPage = ({ params }: { params: { id: string } }) => {
+  const [isEditable, setIsEditable] = useState(false);
+  const [selectedImage, setSelectedImage] = useState<string | null>(null);
+  const [title, setTitle] = useState('N/A');
+  const [slug, setSlug] = useState('N/A');
+  const [image, setImage] = useState('');
+  const [author, setAuthor] = useState({ first_name: 'N/A', last_name: 'N/A' });
+  const [publishDate, setPublishDate] = useState('N/A');
+  const [tags, setTags] = useState<string[]>([]);
```
Comment (edichoska) [inline app/content-panel/blogs/[id]/page.tsx:15]: You can reduce the number of state variables and combine them in one state object called blogDetails

Comment (edichoska) [reply app/content-panel/blogs/[id]/page.tsx:15]: You can maybe keep the image state separate if you want to show a preview of the uploaded image

File: app/content-panel/blogs/[id]/page.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    setAuthor((prev) => ({ ...prev, [name]: value }));
+  };
+
+  const handleDateChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setPublishDate(event.target.value);
+  };
+
+  const handleTagsChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTags(event.target.value.split(',').map((tag) => tag.trim()));
+  };
```
Comment (edichoska) [inline app/content-panel/blogs/[id]/page.tsx:70]: if we have one single state object, you can remove all this onChange handlers and create one that would handle the updates

File: app/content-panel/blogs/[id]/page.tsx (hunk 5 above)
Comment (mitkapanarin) [reply app/content-panel/blogs/[id]/page.tsx:15]: single state object created

File: app/content-panel/blogs/[id]/page.tsx (hunk 2 above)
Comment (mitkapanarin) [reply app/content-panel/blogs/[id]/page.tsx:11]: Since the slug is used for routing and is not intended for user modification, I should remove the slug field and state entirely from the BlogDetailsCard component,
and should not display the slug in the UI ?

Comment (edichoska) [reply app/content-panel/blogs/[id]/page.tsx]: @mitkapanarin yes, thats correct.

File: app/content-panel/blogs/[id]/page.tsx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        image: image || '',
+        author: author || { first_name: 'N/A', last_name: 'N/A' },
+        publishDate: publishDate || 'N/A',
+        tags: tags || [],
+      });
+    };
+    fetchData();
+  }, [params.id]);
+
+  const handleEditClick = () => {
```
Comment (stoimenov95) [inline app/content-panel/blogs/[id]/page.tsx:53]: Just a tip when you work on this ticket next, please investigate if memoization is needed.

File: components/module-components/create-blogs/BlogListView.tsx [hunk 8]
```diff
@@ -43,8 +45,9 @@ const BlogListView = () => {
           throw new Error(`${response.status} ${response.statusText}`);
         }
         const result = await response.json();
+        console.log('result', result);
```
Comment (stoimenov95) [inline components/module-components/create-blogs/BlogListView.tsx:48]: Please remove unnecessary logs.

**Review by stoimenov95 — CHANGES_REQUESTED** [review]
@mitkapanarin great job overall, after the initial testing, I can only mention that we may need to align the design just a bit - update the inputs to have the same design as the first one (size, border, border-radius, etc.).

File: app/content-panel/blogs/[id]/page.tsx (hunk 7 above)
Comment (mitkapanarin) [reply app/content-panel/blogs/[id]/page.tsx:74]: I ll do that

File: components/module-components/create-blogs/BlogListView.tsx (hunk 8 above)
Comment (mitkapanarin) [reply components/module-components/create-blogs/BlogListView.tsx:48]: removed

File: app/content-panel/blogs/[id]/edit/page.tsx [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+  last_name: string;
+}
+
+interface BlogDetails {
+  title: string;
+  content: string;
+  image: string;
+  author: Author;
+  publishDate: string;
+  tags: string[];
```
Comment (edichoska) [inline app/content-panel/blogs/[id]/edit/page.tsx:19]: We dont need to create another page to be able to edit the blog post. It will remain the same page. On Click on Edit all inputs will become enabled and can be edited.

Comment (edichoska) [reply app/content-panel/blogs/[id]/edit/page.tsx:19]: Additionally this should be on another branch after we merge this one, which is made for the ticket: view individual blog page.
You can revert this changes with git. Look into how you can see the last stable commit(the one before these changes with edit), and revert the branch head to that commit :D
Let me know if you need any help doing that.

Comment (mitkapanarin) [reply app/content-panel/blogs/[id]/edit/page.tsx:19]: > We dont need to create another page to be able to edit the blog post. It will remain the same page. On Click on Edit all inputs will become enabled and can be edited.

My previous code met these requirements, but I considered creating a separate edit page to enable redirection from the Blog List page when clicking the edit button. Without a separate page, I'm unsure how to access the edit mode directly.
Screenshot 2024-08-29 231108



What do you think, should I bring back my old code, or its ok to stay like it is now ?

Comment (mitkapanarin) [reply app/content-panel/blogs/[id]/edit/page.tsx:19]: Or should I revert the changes and move the current code to a new branch called edit-blog, or is the separate EditBlog page unnecessary altogether?

Comment (edichoska) [reply app/content-panel/blogs/[id]/edit/page.tsx:19]: The problem here is:
You have `blogs page`and `individual blog page`. On Click on **Edit** on the `blogs page` you should be redirected to the `individual blog page`, but the `individual blog page` should be in **_edit mode_**.
How can these two pages have that state available to both of them?

Comment (edichoska) [reply app/content-panel/blogs/[id]/edit/page.tsx:19]: The EditBlog is unnecessary. @mitkapanarin

File: app/content-panel/blogs/[id]/page.tsx [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const response = await fetch(
+          `${process.env.NEXT_PUBLIC_API_BASE_URL}/blog-posts/${params.id}`
+        );
+        if (!response.ok) {
+          throw new Error(`HTTP error! Status: ${response.status}`);
+        }
+        const data = await response.json();
```
Comment (edichoska) [inline app/content-panel/blogs/[id]/page.tsx:44]: please look into api/queries. There you can find how we are fetching data for this project. We are using react-query.
Remove this block of code and create another file there, or see if you can use api/queries/blogs/getBlogs.ts

File: app/content-panel/blogs/[id]/page.tsx [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+      } catch (fetchError) {
+        setHasError(true);
+        setBlogDetails({
+          title: 'Error fetching data',
+          content: '',
+          image: '',
+          author: { first_name: 'N/A', last_name: 'N/A' },
+          publishDate: 'N/A',
+          tags: [],
+        });
```
Comment (edichoska) [inline app/content-panel/blogs/[id]/page.tsx:63]: don't set data in catch block.

File: app/content-panel/blogs/[id]/page.tsx [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+      setUploadError(null);
+
+      const reader = new FileReader();
+      reader.onload = () => {
+        setSelectedImage(reader.result as string);
+      };
+      reader.readAsDataURL(file);
+    }
+  };
```
Comment (edichoska) [inline app/content-panel/blogs/[id]/page.tsx:108]: it would be nice in the edit ticket to implement something like react-dropzone, so we can drag and drop images. 
Just a note for the other ticket, here seems to be fine.

**Review by edichoska — CHANGES_REQUESTED** [review]
@mitkapanarin Everything looks good to me.
I would also like to be able to click on the table row, and that would lead me to the individual blog page. (same functionality as on View button)
Once that is done, we can merge this.

File: components/reusable-components/reusable-table/TableRowComponent.tsx [hunk 13]
```diff
@@ -10,46 +10,83 @@ interface TableRowComponentProps<T extends { id: string }> {
...
+          <button
+            type="button"
+            className={style.actionButtons}
+            onClick={handleActionClick}
+            onKeyDown={handleKeyDown}
+            tabIndex={0} // Make button focusable
+            aria-haspopup="true" // Indicate a dropdown menu
+          >
+            {renderActions(data)}
+          </button>
```
Comment (edichoska) [inline components/reusable-components/reusable-table/TableRowComponent.tsx:89]: You have two nested buttons here. 
Screenshot 2024-09-08 at 15 20 39
Please revert the changes you did in the ActionDropdown. You can pass onClick function and also styles if you need to overwrite the default styles we already have.

File: components/reusable-components/reusable-table/TableRowComponent.tsx [hunk 14]
```diff
@@ -10,46 +10,83 @@ interface TableRowComponentProps<T extends { id: string }> {
   onCheckboxChange: (id: string) => void;
   showCheckbox?: boolean;
   renderActions?: (item: T) => React.ReactNode;
-  renderActionsDropdown?: ReactNode;
+  renderActionsDropdown?: React.ReactNode;
+  onClick?: (id: string) => void;
```
Comment (edichoska) [inline components/reusable-components/reusable-table/TableRowComponent.tsx:14]: onClick should not be an optional prop

File: components/reusable-components/reusable-table/TableRowComponent.tsx [hunk 15]
```diff
@@ -10,46 +10,83 @@ interface TableRowComponentProps<T extends { id: string }> {
...
 }: TableRowComponentProps<T>) => {
   const handleCheckboxChange = () => {
     onCheckboxChange(data.id);
   };
 
+  const handleRowClick = () => {
+    if (onClick) {
+      onClick(data.id);
+    }
+  };
```
Comment (edichoska) [inline components/reusable-components/reusable-table/TableRowComponent.tsx:35]: Since onClick is not an optional prop please review this function.

File: components/reusable-components/reusable-table/TableRowComponent.tsx [hunk 16]
```diff
@@ -10,46 +10,83 @@ interface TableRowComponentProps<T extends { id: string }> {
...
+  const handleActionClick = (event: React.MouseEvent) => {
+    event.stopPropagation(); // Prevent row click from being triggered
+  };
+
+  const handleKeyDown = (event: React.KeyboardEvent) => {
+    if (event.key === 'Enter' || event.key === ' ') {
+      handleActionClick(event as any);
+    }
+  };
+
```
Comment (edichoska) [inline components/reusable-components/reusable-table/TableRowComponent.tsx:46]: Do we have need for both of these?

File: components/reusable-components/blogDetails-card/BlogDetailsCard.tsx [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+        )}
+      </div>
+
+      <div className={styles.contentSection}>
+        <label htmlFor="contentEditor">Content:</label>
+        <TiptapEditor
+          content={content}
+          editable={isEditable}
+          onChange={(editorContent) =>
+            onChange({ target: { name: 'content', value: editorContent } } as any)
```
Comment (edichoska) [inline components/reusable-components/blogDetails-card/BlogDetailsCard.tsx:82]: Since we are using Typescript, lets try not to leave any "anys" in the code.

File: app/content-panel/blogs/[id]/page.tsx [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+    title: '',
+    image: '',
+    content: '',
+    author: {
+      first_name: '',
+      last_name: '',
+    },
+    publishDate: '',
+    tags: [] as string[],
+  });
```
Comment (edichoska) [inline app/content-panel/blogs/[id]/page.tsx:21]: You should create an interface for the BlogDetailsData and use it to infer types in the useState

**Review by edichoska — CHANGES_REQUESTED** [review]
Please take a look and remove any comments you have left 🙏

File: components/reusable-components/reusable-table/ReusableTable.tsx
Comment (edichoska) [inline components/reusable-components/reusable-table/ReusableTable.tsx:1]: We are aiming to make this component reusable. So it would be a good approach here to remove the ` handleView` function. And pass it to the ReusableTable component as a prop. Then we will pass that function to the Table Row.

**Review by edichoska — APPROVED** [review]
LGTM 🚀

**Review by edichoska — APPROVED** [review]

**Review by stoimenov95 — APPROVED** [review]
LGTM
